site/src/Librecores/ProjectRepoBundle/RepoCrawler/GitRepoCrawler.php

```diff
@@ -164,4 +179,124 @@ public function getLicenseTextSafeHtml(): ?string
 
         return $sanitizedHtml;
     }
+
+    /**
+     * {@inheritDoc}
+     * @see RepoCrawler::fetchCommits()
+     */
+    public function fetchCommits(?string $sinceId = null)
+    {
+
+        $args = ['log', '--reverse', '--format=%H|%aN|%aE|%aD', '--shortstat'];
+
+        if (null !== $sinceId) {
+            // we don't need escapeshellargs here
+            // it is performed internally by ProcessBuilder
+            $args[] = $sinceId;
+            $args[] = '...';
+        }
+
+        $cwd = $this->getRepoClonePath();
+
+        $this->logger->info("Fetching commits in $cwd");
+
+        $output = $this->executor->exec('git', $args, [
+            'cwd' => $cwd,
+            'timeout' => self::TIMEOUT_GIT_LOG
+        ]);
+
+        $this->logger->debug("Fetched commits from $cwd");
+        $this->parseCommits($output);
+    }
+
+    /**
+     * {@inheritDoc}
+     * @see RepoCrawler::commitExists()
+     */
+    public function commitExists(string $id) : bool
+    {
+        // Stolen from https://stackoverflow.com/a/13526591
+        $args = [ 'merge-base', '--is-ancestor', $id, 'HEAD' ];
+        $cwd = $this->getRepoClonePath();
+
+        $this->logger->info('Checking commits in ' . $cwd);
+
+        $this->executor->exec('git', $args, [
+            'timeout' => self::TIMEOUT_GIT_LOG,
+            'cwd' => $cwd,
+            'errors' => true
+        ], $code, $errors);
+        switch ($code)
+        {
+            case 0:     // successful, commit exists
+            case 1:     // commit does not exist in branch
+            case 128:   // commit does not exist in repository
+                $this->logger->debug("Checked commits from $cwd");
+                return $code === 0;
+            default:    // anything other than 0 or 1 is error
+                //die($code);
+                throw new \RuntimeException("Unable to fetch commits from $cwd : ".$errors);
+        }
```

Comment:
This switch statement needs refactoring. "// anything other than 0 or 1 is error" doesn't match reality, and "return $code === 0;" together with the fall-through case statements is hard to understand. Spend a line more code to make it readable.